src/scaler/mod.rs

```diff
@@ -39,18 +60,363 @@ pub trait Scaler {
+    }
+
+    /// Handles an incoming event for the given scaler.
+    ///
+    /// This function processes the event and returns a vector of commands to be executed.
+    /// It also manages the expected events list, removing successfully handled events
+    /// and adding new expected events based on the executed commands, and using the notifier
+    /// to send notifications to other scalers running on different wadm instances.
+    ///
+    /// # Arguments
     ///
-    /// Please note that we'd like this to be temporary in the long term
-    async fn backoff(&self, notifier: Sender<String>);
+    /// * `scaler`: A reference to the `ScalerWithEvents` struct which represents the scaler with events.
+    /// * `event`: A reference to the `Event` struct which represents the incoming event to be handled.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<Vec<Command>>`: A `Result` containing a vector of `Command` structs if successful,
+    ///   or an error of type `anyhow::Error` if any error occurs while processing the event.
+    #[instrument(level = "trace", skip_all, fields(scaler_id = %self.id()))]
+    async fn handle_event_internal(&self, event: &Event) -> anyhow::Result<Vec<Command>> {
+        let model_name = &self.model_name;
+        let commands: Vec<Command> = if self.remove_event(event).await? {
+            trace!("Scaler received event that it was expecting");
+            let data = serde_json::to_vec(&Notifications::RemoveExpectedEvent {
+                name: model_name.to_owned(),
+                scaler_id: self.scaler.id().to_owned(),
+                event: event.to_owned().try_into()?,
+            })?;
+            self.notifier
+                .publish(data, Some(&self.notify_subject))
+                .await?;
+
+            // The scaler was expecting this event and it shouldn't respond with commands
+            Vec::with_capacity(0)
+        } else if self.event_count().await > 0 {
+            trace!("Scaler received event but is still expecting events, ignoring");
+            // If a scaler is expecting events still, don't have it handle events. This is effectively
+            // the backoff mechanism within wadm
+            Vec::with_capacity(0)
+        } else {
+            trace!("Scaler is not backing off, handling event");
+            let commands = self.scaler.handle_event(event).await?;
+
+            // Based on the commands, compute the events that we expect to see for this scaler. The scaler
+            // will then ignore incoming events until all of the expected events have been received.
+            let expected_events = commands
+                .iter()
+                .filter_map(|cmd| cmd.corresponding_event(model_name));
+
+            // Only let other scalers know if we generated commands to take
+            // NOTE(brooksmtownsend): I tried to make the events iterator peekable and look at that instead
+            // but that resulted in a "higher ranked lifetime error". Weird.
```

Comment:
Could link to the rust issue with a TODO to revisit once that's fixed?

Reply:
I just removed this for now (cc @brooksmtownsend), I wasn't sure what he was trying to do with peekable. Can add back in later